File: socialview-commons/src/main/java/com/hendraanggrian/widget/socialview/commons/MentionAdapter.java
package com.hendraanggrian.widget.socialview.commons;

import android.content.Context;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.hendraanggrian.picasso.commons.transformation.Transformations;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.RequestCreator;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class MentionAdapter extends SuggestionAdapter<Mention> {

    @DrawableRes private final int defaultAvatar;

    public MentionAdapter(@NonNull Context context) {
        this(context, R.drawable.ic_placeholder_mention);
    }

    public MentionAdapter(@NonNull Context context, @DrawableRes int defaultAvatar) {
        super(context, R.layout.item_mention, R.id.textview_mention_username);
        this.defaultAvatar = defaultAvatar;
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        final ViewHolder holder;
        if (convertView == null) {
            convertView = LayoutInflater.from(getContext()).inflate(R.layout.item_mention, parent, false);
            holder = new ViewHolder(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        final Mention item = getItem(position);
        if (item != null) {
            holder.textViewUsername.setText(item.getUsername());

            final RequestCreator request;
            if (item.getAvatar() == null)
                request = Picasso.with(getContext())
                        .load(defaultAvatar);
            else if (item.getAvatar() instanceof Integer)
                request = Picasso.with(getContext())
                        .load((int) item.getAvatar());
            else if (item.getAvatar() instanceof String)
                request = Picasso.with(getContext())
                        .load((String) item.getAvatar())
                        .placeholder(defaultAvatar)
                        .error(defaultAvatar);
            else
                throw new RuntimeException("Mentionable avatar can only be String url and int resource.");
            request.transform(Transformations.circle())
                    .fit()
                    .into(holder.imageView);

            if (!TextUtils.isEmpty(item.getDisplayname())) {
                holder.textViewDisplayname.setText(item.getDisplayname());
                holder.textViewDisplayname.setVisibility(View.VISIBLE);
            } else {
                holder.textViewDisplayname.setVisibility(View.GONE);
            }
        }
        return convertView;
    }

    @NonNull
    @Override
    public SuggestionFilter initializeFilter() {
        return new SuggestionFilter() {
            @Override
            public String getString(Mention item) {
                return item.getUsername();
            }
        };
    }

    private static class ViewHolder {
        private final ImageView imageView;
        private final TextView textViewUsername, textViewDisplayname;

        private ViewHolder(@NonNull View view) {
            imageView = (ImageView) view.findViewById(R.id.imageview_mention_username);
            textViewUsername = (TextView) view.findViewById(R.id.textview_mention_username);
            textViewDisplayname = (TextView) view.findViewById(R.id.textview_mention_displayname);
        }
    }
}

File: socialview-commons/src/com/hendraanggrian/widget/MentionAdapter.kt
package com.hendraanggrian.widget

import android.content.Context
import android.net.Uri
import android.support.annotation.DrawableRes
import android.view.LayoutInflater.from
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.hendraanggrian.pikasso.circle
import com.hendraanggrian.pikasso.picasso
import com.hendraanggrian.pikasso.toProgressTarget
import com.hendraanggrian.socialview.Mention
import com.hendraanggrian.socialview.commons.R
import java.io.File

/**
 * Default adapter for displaying mention in [SocialAutoCompleteTextView].
 * Note that this adapter is completely optional, any adapter extending
 * [android.widget.ArrayAdapter] can be attached to [SocialAutoCompleteTextView].
 */
class MentionAdapter @JvmOverloads constructor(
    context: Context,
    @DrawableRes private val defaultAvatar: Int = R.drawable.socialview_ic_mention_placeholder
) : SocialAdapter<Mention>(
    context,
    R.layout.socialview_layout_mention,
    R.id.socialview_mention_username
) {

    override fun Mention.convertToString(): String = username

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        var _convertView = convertView
        when (_convertView) {
            null -> {
                _convertView = from(context)
                    .inflate(R.layout.socialview_layout_mention, parent, false)
                holder = ViewHolder(_convertView!!)
                _convertView.tag = holder
            }
            else -> holder = _convertView.tag as ViewHolder
        }
        getItem(position)?.let { mention ->
            when (mention.avatar) {
                null -> picasso.load(defaultAvatar)
                is Int -> picasso.load(mention.avatar)
                is String -> picasso.load(mention.avatar)
                is Uri -> picasso.load(mention.avatar)
                is File -> picasso.load(mention.avatar)
                else -> error("Unsupported avatar type. See Mention.kt for more.")
            }.error(defaultAvatar)
                .circle()
                .into(holder.avatarView.toProgressTarget(
                    context.resources.getDimensionPixelSize(
                        R.dimen.socialview_mention_progress_bar)))
            holder.usernameView.text = mention.username
            if (!mention.displayname.isNullOrEmpty()) {
                holder.displaynameView.text = mention.displayname
            }
        }
        return _convertView
    }

    private class ViewHolder(itemView: View) {
        val avatarView: ImageView = itemView.findViewById(R.id.socialview_mention_avatar)
        val usernameView: TextView = itemView.findViewById(R.id.socialview_mention_username)
        val displaynameView: TextView = itemView.findViewById(R.id.socialview_mention_displayname)
    }
}